bip-txhash.md

```diff
@@ -0,0 +1,269 @@
+  * `TXFS_INPUTS_TEMPLATE = TXFS_INPUTS_SEQUENCES | TXFS_INPUTS_SCRIPTSIGS | TXFS_INPUTS_PREV_VALUES | TXFS_INPUTS_TAPROOT_ANNEXES`
+  * `TXFS_OUTPUTS_ALL = TXFS_OUTPUTS_SCRIPTPUBKEYS | TXFS_OUTPUTS_VALUES`
+
+For both inputs and then outputs, do the following:
+
+* If the "in/outputs" field is set to 1, another additional byte is expected:
+  * The highest bit (`TXFS_INOUT_NUMBER`) indicates whether the "number of in-/outputs"
+    should be committed to.
+  * For the remaining bits, there are three exceptional values:
+    * 0x00 (`TXFS_INOUT_SELECTION_NONE`) means "no in/outputs"
+      (hence only the number of them as `0x80` (`TXFS_INOUT_NUMBER`)).
+    * `0x40` (`TXFS_INOUT_SELECTION_CURRENT`) means "select only the in/output of the current input index"
+      (it is invalid when current index exceeds number of outputs).
+    * `0x3f` (`TXFS_INOUT_SELECTION_ALL`) means "select all in/outputs".
+
+  * The second highest bit (`TXFS_INOUT_SELECTION_MODE`) is the "specification mode":
+    * Set to 0 it means "leading mode".
+    * Set to 1 it means "individual mode".
+  * In "leading mode", the third highest bit (`TXFS_INOUT_SELECTION_SIZE`) is
+    used to indicate the "count size", i.e. the number of bytes will be used to
+    represent the number of in/output.
+    * With "index size" set to 0, the remaining lowest 5 bits of the first byte will
+      be interpreted as the number of leading in/outputs to select.
+    * With "index size" set to 1, the remaining lowest 5 bits of the first byte together with the
+      8 bits of the next byte will be interpreted as the number of leading in/outputs to select.
+  * In "individual mode", the remaining lowest 6 bits of the first byte will be
+    interpreted as `n`, the number of individual in/outputs to select. For each
+    individual input, (at least) one byte is expected, of this byte. The
+    highest bit is used to indicate "absolute or relative" indices.
+    * If the highest bit is set to 0, it is an absolute index. The second
+      highest bit is used to indicate the amount of bytes are used to represent
+      the index.
+      * If the second-highest bit is 0, the remaining 6 bits represent the index to be selected.
+      * If the second-highest bit is 1, the remaining 6 bits, together with the 8 bits of the next
+        byte, represent the index to be selected.
+    * If the highest bit is set to 1, it is a relative index. The second highest bit is used to
+      indicate the sign of the index.
+      * If the second-highest bit is set to 0, the remaining 6 bits represent the positive relative
+        index to be selected.
+      * If the second-highest bit is set to 1, the remaining 6 bits represent the negative relative
+        index to be selected.
+
+Effectively, this allows a user to select
+* all in/outputs
+* the current input index
+* the leading in/outputs up to 8192
+* up to 64 individually selected in/outputs
```

Comment:
Does this allow everlasting covenants (coins locked forever in a fixed set of addresses)? It is not clear from the text of BIP if this is possible and an intended use-case. IIUC, it is impossible because of chicken-and-egg problem: output script has to include a hash of itself to make an everlasting covenant which is impossible. But it is better to clarify this in BIP text explicitly, including a formal proof of why an everlasting covenant is impossible or (if it is possible) discuss use-cases and consequences.

Reply:
Yeah, this is not possible AFAIK. But I wouldn't be comfortable making that claim, if you see what people can do. Especially if we get OP_CAT and OP_TWEAKADD.